tensornetwork/ncon_interface.py

```diff
@@ -23,143 +23,256 @@
 
 _CACHED_JITTED_NCONS = {}
 
-
-def _get_cont_out_labels(network_structure: Sequence[Sequence]) -> Any:
+def _get_cont_out_batch_labels(
+    network_structure: Sequence[Sequence[Union[int, str]]]) -> Any:
   """
-  Compute the contracted and free labels of `network_structure`.
-  Contracted labels are labels appearing more than once,
-  free labels are labels appearing exactly once.
+  Compute the contracted, free and batch labels of `network_structure`, 
+  using the following rules:
+  * Any negative number-type element and any hyphen-prepended str-type 
+    element are considered output labels.
+  * Any positive number-type element and any non-hyphen-prepended 
+    str-type element are considered contracted labels.
+  * Any negative number-type element appearing more than once, any 
+    hyphen-prepended str-type element appearing more than once, 
+    any positive element appearing exactly once and 
```

Comment:
My personal preference would be to throw an error for string labels containing anything but alphanumeric characters, on the grounds that an index like "JE!@//$" was probably a mistake. Even then, it's going to be very confusing trying to decide whether "Jeffrey0" or "jeff0rey" go first. But I remember some disagreement on this point.

Reply:
The first that is easy enough to fix and seems reasonable. I think using ascii ordering is fine if we tell the users